Add backwards compatibility for the scraper

exitflynn opened the pull request:
Aims to add backwards compatibility for the base scraper post-rewrite from #7077 so that we can go through a deprecation period for the breaking changes.

Addressing @dstansby's [suggestions](https://github.com/sunpy/sunpy/pull/7227#pullrequestreview-1682939171) on how we can approach this, so far this PR pretty much addresses all the points in terms of the Scraper design, namely the PR:

  - Makes `pattern` optional (ie. , pattern=None)
  - Adds a new keyword-only argument, `format`, to take the new format
  - Raises an error if both pattern and format are provided
  - Uses old behaviour, and raises a deprecation warning saying to use new format=... instead if only `pattern` passed
  - Uses new behaviour if only `format` passed.

Outside of changing `sunpy.net.scraper` I've done the following:
- Added back the tests for the original scraper with the ones I wrote (distinguished by a `new_format` suffix) and they all should be passing now.
- During the rewrite, I had previously moved out `check_timerange` from `scraper` to `scraper_utils` but I moved it back in so long as the old scraper is still around since I felt it'd get a bit too complex otherwise (for example, it'd require internal scraper functions that were removed in the rewrite, etc).
- Updated hopefully all the relevant docstring examples and the base dataretriever clients' code. I am _not_ sure of all the tests passing, i'm curious of the CI's verdict on that rn :P
- I've also updated the `extending_fido.rst` doc about `pattern` vs `format` and have tried to improve some of the documentation I wrote previously explaining the Scraper algorithm.
- *also renamed `timepattern` to ~~`dt_pattern`~~ `datetime_pattern` since it's also appearing in docs rn, felt it more fitting.

The really long file for the scraper tests right now make me doubt a little if maybe I should've skimmed over some tests when adding backwards compatibility, I'd love to know if I've been on the right path with this so far :cowboy_hat_face:

TODO:
- [x] We need to add a test that creates an old style scraper client (basically use one of the old clients) and make sure that it raises the warning correctly and still works. (Added back nearly all the old teststhat expect and catch the warning  and renamed new ones to have a `new_format` suffix)
- [x] Write a migration guide in the developer docs and link to it in the message.

nabobalis commented:
Thanks, I will try to get some time to review this next week.

samaloney commented on sunpy/net/scraper_utils.py:
```diff
@@ -77,33 +86,66 @@ def date_floor(date, timestep):
     return datetime(*time_tup)
 
-def check_timerange(pattern, url, timerange):
+def extractDateURL(self, url):
```
```suggestion
def extract_date(self, url):
```

samaloney commented on sunpy/net/scraper_utils.py:
```diff
@@ -77,33 +86,66 @@ def date_floor(date, timestep):
-    tr = get_timerange_from_exdict(exdict)
-    return tr.intersects(timerange)
+    # remove the user and passwd from files if there:
+    url = url.replace("anonymous:data@sunpy.org@", "")
```
Can a user supply different credentials and if so would this break?

exitflynn replied:
i cant think of an example that'd hit this but i don't think it should break, this was most likely originally added since the current implementation [overrides ftp filelist-ing with these credentials]( https://github.com/sunpy/sunpy/blob/scraper_rewrite/sunpy/net/scraper.py#L207) and this function internally gets called later on.

Generalising it would be a simple regex change though, let me know if it seems like a good change to have.

However I'm a bit skeptic about this check ever passing since after looking up FTP URL structure the second `@` shouldn't be there?

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -27,29 +27,42 @@ A new "scraper" client inherits from `~sunpy.net.dataretriever.client.GenericCli
-* A class attribute ``pattern``; this is a string used to match the URLs supported by the client and extract metadata from the matched URLs.
-  The time and other metadata attributes for extraction are written using the `~sunpy.extern.parse.parse` format, using double curly-brackets so to differentiate them from ``kwargs`` parameters which are written in single curly-brackets.
+* A class attribute ``pattern``; this is a string used to match all URLs supported by the client and extract necessary metadata from the matched URLs.
+  The time and other metadata attributes for extraction are written within double curly-braces ``{{}}`` in a `parse` format. Regular placeholders for Python format strings can still be included via single curly braces ``{}`` with their respective parameters passed as the ``kwargs``. An example of how such a pattern looks like is given in the algorithm explanation below.
```
```suggestion
  The time and other metadata attributes for extraction are written within double curly-braces ``{{}}`` in a `parse` format.
  Regular placeholders for Python format strings can still be included via single curly braces ``{}`` with their respective parameters passed as the ``kwargs``.
  An example of how such a pattern looks like is given in the algorithm explanation below.
```

exitflynn commented:
Thanks for your reviews! 

> This looks great, nice work. I know it came up before but do we have a migration guide or section in the docs explaining how to eventually move to the new format syntax.

A guide for writing clients with the newer format is added in the all-things-scraper-clients / extending-Fido [topic-guide](https://github.com/exitflynn/sunpy/blob/scraper_rewrite/docs/topic_guide/extending_fido.rst) but it's not explicitly about migrating yet. Does what's there work or should I try and address migration more?

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -27,29 +27,43 @@ A new "scraper" client inherits from `~sunpy.net.dataretriever.client.GenericCli
+  Regular placeholders for Python format strings can still be included via single curly braces ``{}`` with their respective parameters passed as the ``kwargs``.
+  An example of how such a pattern looks like is given in the algorithm explanation below.
+
+Note: The scraper has supported regex-based patterns for a long time, which is what it will still expect by default. However that is in the process of being replaced with parse-style patterns in the future versions of the package. To use the newer parse-style patterns, it's currently required to pass it as the ``format`` argument to the scraper.
```
```suggestion
Note: The scraper has supported regex-based patterns for a long time, which is what it will still expect by default.
However that is in the process of being replaced with parse-style patterns in the future versions of the package.
To use the newer parse-style patterns, it's currently required to pass it as the ``format`` argument to the scraper.
```

nabobalis replied:
We want to enable the new format by default and raise a warning if its the old way.

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -27,29 +27,43 @@ A new "scraper" client inherits from `~sunpy.net.dataretriever.client.GenericCli
+For a more verbose in-depth explanation on how this is accomplished internally:
 
-1. It takes as input a generalised ``pattern`` of how a desired filepath looks like, following the ``parse`` format. A version of the pattern following the datetime format is also generated, called the ``timepattern``.
+1. A Scraper object takes as input the generalised ``pattern`` of how a desired filepath looks like, in the ``parse`` format. Upon initialisation, a version of the pattern following the datetime format is also internally generated, called the ``dt_pattern``.
```
```suggestion
1. The Scraper takes as input the generalised ``pattern`` of how a desired file path looks like, in the ``parse`` format.
   Upon initialisation, a version of the pattern following the datetime format is also internally generated, called the ``dt_pattern``.
```

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -27,29 +27,43 @@ A new "scraper" client inherits from `~sunpy.net.dataretriever.client.GenericCli
     ...            '{instrument}_lv1_{{year:4d}}{{month:2d}}{{day:2d}}_{{hour:2d}}{{minute:2d}}{{second:2d}}.fits')
-    >>> s = Scraper(pattern, instrument='swap')
+    >>> s = Scraper(format=pattern, instrument='swap')
+    >>> s.dt_pattern
```
Thoughts on renaming this to "datetiime_pattern"?

exitflynn replied:
i think that works better, yeah

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -27,29 +27,43 @@ A new "scraper" client inherits from `~sunpy.net.dataretriever.client.GenericCli
+    'http://proba2.oma.be/swap/data/bsd/%Y/%m/%d/swap_lv1_%Y%m%d_%H%M%S.fits'
 
-2. The smallest unit of time / time-step for that directory pattern (the full timepattern minus the filename at the end) is then detected by using :meth:`~sunpy.net.scraper_utils.extract_timestep`.
+The smallest unit of time / time-step for that directory-pattern (the full ``dt_pattern`` except the filename at the end) is then internally detected from ``dt_pattern`` by using the :meth:`~sunpy.net.scraper_utils.extract_timestep` util.
```
```suggestion
The smallest unit of time / time-step for that directory-pattern (the full ``dt_pattern`` except the filename at the end) is then internally detected from ``dt_pattern`` by using :meth:`~sunpy.net.scraper_utils.extract_timestep`.
```

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -60,9 +74,9 @@ Each such client relies on the `~sunpy.net.scraper.Scraper` to be able to query
-4. The location given by the filled pattern is visited and a list of files at the location is obtained. This is handled differently depending on whether the pattern is a web URL or a ``file://`` or an ``ftp://`` path in the :meth:`~sunpy.net.scraper.Scraper.filelist` method.
-5. The name of each file present is then examined to determine if it matches the remaining portion of the pattern using :meth:`~sunpy.extern.parse.parse`.
-6. Each such file is then checked for lying in the intended timerange using the :meth:`~sunpy.net.scraper_utils.check_timerange` method which in turn uses :meth:`sunpy.net.scraper_utils.get_timerange_from_exdict` to get the covered timerange for each file. The files that satisfy these conditions are then added to the output.
+2. The location given by the filled pattern is visited and a list of files at the location is obtained. This is handled differently depending on whether the pattern is a web URL or a ``file://`` or an ``ftp://`` path in the :meth:`~sunpy.net.scraper.Scraper.filelist` method.
+3. Each filename is then parsed against the remaining portion of the pattern to determine if it matches.
+4. Each such file is then checked for lying in the intended timerange using the :meth:`~sunpy.net.scraper._check_timerange` method which in turn uses :meth:`sunpy.net.scraper_utils.get_timerange_from_exdict` to get the covered timerange for each file. The files that satisfy these conditions are then added to the output.
```
```suggestion
The location given by the filled pattern is visited and a list of files at the location is obtained.
This is handled differently depending on whether the pattern is a web URL or a ``file://`` or an ``ftp://`` path by :meth:`~sunpy.net.scraper.Scraper.filelist`.

Each filename is then parsed against the remaining portion of the pattern to determine if it matches.
Each such file is then checked for lying in the intended timerange using the :meth:`~sunpy.net.scraper._check_timerange` method which in turn uses :meth:`sunpy.net.scraper_utils.get_timerange_from_exdict` to get the covered timerange for each file.
The files that satisfy these conditions are then added to the output.
```

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -145,18 +159,11 @@ Examples
 Suppose any file of a data archive can be described by this URL ``https://some-domain.com/%Y/%m/%d/satname_{SatelliteNumber}_{Level}_%y%m%d%H%M%S_{any-2-digit-number}.fits``:
 
-``baseurl`` becomes ``r'https://some-domain.com/%Y/%m/%d/satname_(\d){2}_(\d){1}_(\d){12}_(\d){2}\.fits'``.
+The new format ``pattern`` becomes ``r'https://some-domain.com{{year:4d}}/{{month:2d}}{{day:2d}}/satname_{SatelliteNumber:2d}_{Level:1d}_{{year:2d}}{{month:2d}}{{day:2d}}{{hour:2d}}{{minute:2d}}{{second:2d}}_{{:2d}}.fits'``.
+The date-time values and any other metadata attributes that we wish to extract are written within double curly-braces ``{{}}``. These metadata attributes are the desired keys for the returned dictionary and they should match with the ``attr.__name__``.
+Note that parts of such attributes can accordingly be omitted to match parts of the filename which are dynamic but not needed to be extracted. For example, ``{{:2d}}`` is used in the above example to match any 2-digit number in the filename. Similarly ``{{}}`` can be used to match a string of any length starting from its position in the filename.
```
```suggestion
The new format ``pattern`` becomes ``r'https://some-domain.com{{year:4d}}/{{month:2d}}{{day:2d}}/satname_{SatelliteNumber:2d}_{Level:1d}_{{year:2d}}{{month:2d}}{{day:2d}}{{hour:2d}}{{minute:2d}}{{second:2d}}_{{:2d}}.fits'``.
The datetime values and any other metadata attributes that we wish to extract are written within double curly-braces ``{{}}``.
These metadata attributes are the desired keys for the returned dictionary and they should match with the ``attr.__name__``.
Note that parts of such attributes can accordingly be omitted to match parts of the filename which are dynamic but not needed to be extracted.
For example, ``{{:2d}}`` is used in the above example to match any 2-digit number in the filename.
Similarly ``{{}}`` can be used to match a string of any length starting from its position in the filename.
```

nabobalis commented on docs/topic_guide/extending_fido.rst:
```diff
@@ -175,6 +182,7 @@ So the desired key names for returned dictionary should be written in the ``patt
 
         return adict
 
+
```
```suggestion
```

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -40,18 +54,32 @@ class Scraper:
+        A string containing the url with the date encoded as datetime formats,
+        and any other parameter as ``kwargs`` as a string format.
+        This can also be a uri to a local file patterns. Deprecated in favor of `format`. Default is `None`.
+
```
You should remove the line breaks between each parameter.

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -72,24 +102,55 @@ class Scraper:
-            self.now = datetime.now().strftime(self.timepattern)
+    def __init__(self, pattern=None, regex=False, format=None, **kwargs):
+        if pattern is not None and format is None:
+            warnings.warn(
```
I thought we had a helper to raise deprecation warnings for keywords. I can't seem to find it, only for positional arguments.

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -72,24 +102,55 @@ class Scraper:
+            warnings.warn(
+                "Using 'pattern' for the URL format is deprecated. "
+                "Please use 'format' instead.",
+                DeprecationWarning,
```
This should be a sunpydepcreationwaring instead.

exitflynn replied:
hey i can't seem to figure out which of the existing decorators would apply in this case, [here they are for reference](https://github.com/sunpy/sunpy/blob/main/sunpy/util/decorators.py). does one of them fit here or does this call for a new one for deprecating specific arguments?

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -72,24 +102,55 @@ class Scraper:
+            )
+            self.use_old_format = True
+            self.pattern = pattern
+        elif pattern is None and format is not None:
+            self.use_old_format = False
+            self.pattern = format
```
So if both are set, it will use format and ignore the pattern?

I would rather this errored and told users to only format.

exitflynn replied:
i think the else condition afterwards would catch that instead and raise the error?

nabobalis replied:
We should check with a unit test.

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -192,7 +260,7 @@ def filelist(self, timerange):
         elif urlsplit(directories[0]).scheme == "file":
             return self._localfilelist(timerange)
-        elif urlsplit(directories[0]).scheme == "http":
+        elif urlsplit(directories[0]).scheme == "http" or urlsplit(directories[0]).scheme == "https":
```
```suggestion
        elif urlsplit(directories[0]).scheme in ["http", "https"]:
```

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -308,14 +397,131 @@ def _httpfilelist(self, timerange):
+            tr = get_timerange_from_exdict(exdict)
+            return tr.intersects(timerange)
+
+    def _URL_followsPattern(self, url):
```
```suggestion
    def _url_follows_pattern(self, url):
```
Elsewhere as well.

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -308,14 +397,131 @@ def _httpfilelist(self, timerange):
+        """
+        Extracts the date from a particular url following the pattern.
+        """
+        # remove the user and passwd from files if there:
```
```suggestion
        # Remove the user and password if present
```

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -308,14 +397,131 @@ def _httpfilelist(self, timerange):
+            # Substitutes '.' and '_' for '/'.
+            return re.sub(r'\.|_', '/', txt).split('/')
+
+        # create a list of all the blocks in times - assuming they are all
+        # separated with either '.', '_' or '/'.
```
```suggestion
        # Create a list of all the blocks in times
        # assuming they are all separated with either '.', '_' or '/'
```

nabobalis reviewed (dismissed):
Thanks for doing all of this! I left some minor comments. I also rebased the base branch, you might need to rebase.

nabobalis commented:
> but it's not explicitly about migrating yet. Does what's there work or should I try and address migration more?

There should be a short migration guide either in the change log or in the main fido client doc so we can point to it in the deprecation warning.

nabobalis commented:
We also should document what needs to be removed when we want to drop the old code paths

nabobalis commented on sunpy/net/tests/test_scraper.py:
```diff
@@ -1,94 +1,173 @@
 
 def testDirectoryDatePattern():
-    s = Scraper('%Y/%m/%d/%Y%m%d_%H%M%S_59.fit.gz')
+    with pytest.deprecated_call():
```
When you change to sunpydepcreationwarning, I am unsure if this will work. 

I would also like to check the message is the one we expect to be raised.

nabobalis commented on sunpy/net/tests/test_scraper.py:
```diff
@@ -1,94 +1,173 @@
-]
-
 
 def testDirectoryDatePattern():
```
Can we rename these tests to follow the others in sunpy?

`test_directory_date_pattern` and etc.

nabobalis commented on sunpy/net/scraper.py:
```diff
@@ -80,24 +99,77 @@ class Scraper:
 
-    def __init__(self, pattern, regex=False, **kwargs):
-        if regex:
+    def __init__(self, format=None, pattern=None, regex=False, **kwargs):
```
Sorry, I was wrong in my previous comment about this. I will blame covid and not my own incompetence.

We want to keep the API the same way around as it will break all the older clients. 

I haven't thought through what I am about to say but I think we would need to do this:

"pattern", "regex "are the first two arguments to the init, then it will be "format" for the new behavior. 

We check if pattern/regex are passed in, we issue a deprecation warning saying to use format keyword and provide a message with a link to a migration guide.

I would also make the format keyword to be required to be a keyword as a temporary workaround. TO avoid any API ambiguity.

exitflynn replied:
> I would also make the format keyword to be required to be a keyword as a temporary workaround.
ooh do you mean having the format keyword as a required argument? because wouldn't that also break all the older clients?

also wish you a fast and comfy recovery from covid!! and thanks for all the reviews, health issues have also got to me these days, i should be able to address them over the weekend

nabobalis replied:
> > I would also make the format keyword to be required to be a keyword as a temporary workaround.
> 
> do you mean having the format keyword as a required argument? because wouldn't that also break all the older clients?

No. I mean the use of `*` in the signature. 

`def __init__(self, pattern=None, regex=False, *, format=None, **kwargs):`

> also wish you a fast and comfy recovery from covid!! and thanks for all the reviews, health issues have also got to me these days, i should be able to address them all over the weekend.

Make sure to look after yourself!

exitflynn commented:
Thanks for the reviews Nabil, I've added back the `deprecated_renamed_argument` decorator from [astropy](https://github.com/astropy/astropy/blob/main/astropy/utils/decorators.py#L321) to handle the deprecation warnings and addressed pretty much all of the reviews so far. What remains for now is a proper migration guide at the changelog or the docs (or I'm missing something lol) which I should be able to get to very soon!

nabobalis commented:
> Thanks for the reviews Nabil, I've added back the `deprecated_renamed_argument` decorator from [astropy](https://github.com/astropy/astropy/blob/main/astropy/utils/decorators.py?rgh-link-date=2024-06-26T22%3A15%3A52Z#L321) to handle the deprecation warnings and addressed pretty much all of the reviews so far. What remains for now is a proper migration guide at the changelog or the docs (or I'm missing something lol) which I should be able to get to very soon!

Sounds good! I will review this again.

nabobalis commented:
In the meantime I will rebase this PR to fix those conflicts. You will need to reset your local branch when I do this to avoid having to force push again.

nabobalis commented:
TODO:
We need to add a test that creates an old style scraper client (basically use one of the old clients) and make sure that it raises the warning correctly and still works.

nabobalis commented:
I screwed up the rebase @exitflynn, you can force push your local branch to fix my mistake please?

nabobalis commented:
I rebased this, hopefully I got everything.

nabobalis commented:
That online fail is real.

nabobalis commented:
Oh its from helio. Nevermind.

nabobalis commented:
Thanks again @exitflynn
